src/java_bytecode/java_utils.cpp

```diff
@@ -20,3 +22,94 @@ bool java_is_array_type(const typet &type)
     to_struct_type(type).get_tag()),
                     "java::array[");
 }
+
+unsigned java_parameter_count(
+  const std::string method_descriptor,
+  java_parameter_count_whatt what)
+{
+  // Examples (from javap):
+  // descriptor: (ILjava/lang/String;[Ljava/lang/Integer;)I
+  // descriptor: (Ljava/lang/Object;[Ljava/lang/Integer;)I
+  // descriptor: (IF[I)V
+
+  // Base type: B C D F I J S Z
+  // Object type: L "Class name" ;
+  // (Multi-dimension) Array type: [* Type
+
+  unsigned i=0;
+  unsigned num_params=0;
+  unsigned num_slots=0;
+  bool saw_array;
+
+  INVARIANT(method_descriptor.size()>=3, "function always look like ()T");
+  INVARIANT(method_descriptor[0]=='(', "function always look like ()T");
+
+  for(i=1; method_descriptor[i]!=')'; i++)
+  {
+    // we will parse 1 argument in this iteration
+    // all parameter types consume 1 local variable slot except double and long,
+    // which consume 2
+    num_params++;
+    num_slots++;
+    saw_array=false;
+
+    INVARIANT(i<method_descriptor.size(), "health check, closing ) present");
+    if(i>=method_descriptor.size())
+      return num_params;
+
+    // Skip the (multi-dimension) array character, counts as 1 parameter, i will
+    // point to the component type after this
+    if(method_descriptor[i]=='[')
+    {
+      saw_array=true;
+      for(; method_descriptor[i]=='['; i++) {}
+    }
+
+    if(method_descriptor[i]=='L')
+    {
+      // Parse an object reference, skip class name and find the ;
+      for(; method_descriptor[i]!=';'; i++) {}
+    }
+    else
+    {
+      // Parse a base type
+      INVARIANT(
+        method_descriptor[i]=='B' ||
+        method_descriptor[i]=='C' ||
+        method_descriptor[i]=='D' ||
+        method_descriptor[i]=='F' ||
+        method_descriptor[i]=='I' ||
+        method_descriptor[i]=='J' ||
+        method_descriptor[i]=='S' ||
+        method_descriptor[i]=='Z',
+        "the base type is an expected character");
+
+      // parameters of type double and long consume 2 variable slots
+      if((method_descriptor[i]=='D' || method_descriptor[i]=='J')
+            && !saw_array)
+        num_slots++;
+    }
+  }
+
+  // return the requested number
+  switch(what)
+  {
+  case java_parameter_count_whatt::PARAMETERS:
+    return num_params;
+  case java_parameter_count_whatt::VARIABLE_SLOTS:
+    return num_slots;
+  default:
+    UNREACHABLE;
+    return 0;
+  }
+}
```

Comment:
this seems to duplicate what already exists to get a java type from a descriptor (`java_type_from_string`), but with the constraint on function types